### ISSUE TITLE
error in retrieving remote data with setting default

### ISSUE THREAD

**Issue by ade72** [select2/select2#3011]
Retrieving remote data works fine using the standard data adapter. Problem is: When I set ONE default SELECTED option before calling ('#selectbox').select2(); this option is displayed regulary in the select2 as the selected item (one of one). When the user then types in some search criteria, the ajax routine seems to search only in the already populated list of this select-element. Because I only set ONE option (to set a default value in this field), the data in my database seems not to be searched....
The ajax-routine works fine when I do not set the default option.
Did I miss something?

**Comment by ade72** [comment]
Any suggestions? For me this is a show-stopper if I cannot set a default...

**Comment by kevin-brown** [comment]
@ade72 Can you provide an example (or some code we can look at) that can be used to help trace the issue?

Our AJAX example appears to be working fine in the docs, so it's diffcult to help track down the issue without anything to work from.

https://select2.github.io/examples.html#data-ajax

**Comment by ade72** [comment]
$('#' + select2Id).select2(
            {

                width             : '100%',
                placeholder       : 'PLZ oder Stadt',
                cache             : false,
                minimumInputLength: minimumInputLength,
                ajax              : {
                    url           : "/ajax/getCitiesFromPlzOrStadt",
                    dataType      : 'json',
                    delay         : 250,
                    data          : function (params) {

                        return {
                            plzorstadt: params.term
                        };
                    },
                    processResults: function (data) {

                        return {
                            results: data
                        };
                    }
                }

            }
        );


This ajax method works fine and retrieves all places / cities for a given postcode. No problem.
The problem is: If I want to set a DEFAULT in select2, I do it by adding ONE <option value='12345' selected='selected'>Munich</option> line.
If I do so, this is used as a default. So so far no problem.
BUT: When I enter then some chars into the search field, I get NO results....

**Comment by ade72** [comment]
Did I got that right that I have to add ONE <option> line with the DEFAULT value to set though I use select2 for remote data using ajax?

**Comment by ade72** [comment]
Or: How can I set ONE DEFAULT value? In select2 3.5 I used:

```javascript
,
        //        initSelection     : function (element, callback) {
        //
        //            //find element with id that is defined in input value
        //            var locationId = element.val();
        //            $.ajax('/ajax/getLocationFromId', {
        //                data    : {
        //                    locationId: locationId
        //                },
        //                dataType: 'json'
        //            }).done(function (data) {
        //                var modData = {id: locationId, text: data};
        //                callback(modData);
        //            });
        //        },

```

...and it worked fine...

**Comment by kevin-brown** [comment]
Can you check in your developer tools if any errors are being triggered by Select2?

> BUT: When I enter then some chars into the search field, I get NO results.... it seems that select2 uses only this ONE option and not the ajax-call to the database....

Can you confirm that Select2 is making the AJAX call through your browser's developer tools?

> Did I got that right that I have to add ONE line with the DEFAULT value to set though I use select2 for remote data using ajax?

Correct, this should be how you set a default value for AJAX-based selects.

**Comment by ade72** [comment]
Hmmm... good point.
1) no errors are triggered.
2) if I do NOT add this option selected line to my code, the ajax event is triggered, and I see results.
3) if I add the option selected line, the ajax event is not triggered.... hmmm...

maybe a mistake in my code? but the javascript is called anyway...

**Comment by ade72** [comment]
The main difference is, if I set a DEFAULT value for a static options array or an options array that is retrieved "remote" through an ajax call. Setting a default value for a static options array works fine using the method above...

**Comment by ade72** [comment]
3) if I add the option selected line, the ajax event is not triggered....
maybe that line above was not correct... I mean: My developer tools dont show me that the php is called through the ajax-call from javascript.... on the other side: when I type something in select2, it gives me response, that there is NO result... I am stucked...

**Comment by ade72** [comment]
I turned my code round and round, but it only shows this ONE result; that option that I took as the default preset....

**Comment by kevin-brown** [comment]
Lets see if I can narrow this down enough to figure out the issue...

Can you expand upon the issue you are facing by adding a few more pieces of information to your bug report?

- What browser(s) and Operating System have you tested with?
- Does the bug happen consistently across all tested browsers?
- What version of jQuery are you using?
- Are you using Select2 with other plugins?

The [contributing guide][cg] contains a useful section with a few questions which can help us track down the bug and speed up the process of finding a fix for it.  The creation of [an isolated test case][test-case] would also be useful.

[cg]: https://github.com/ivaynberg/select2/blob/master/CONTRIBUTING.md#reporting-bugs-with-select2
[test-case]: http://css-tricks.com/6263-reduced-test-cases/

**Comment by ade72** [comment]
Kevin, I am ashamed to say, but glad that I have found, the fault was on my side. I have written a wrapper php class for select2 3.5, that I used to switch between several kinds of select2 controls, like time controls, static and dynamic controls... fact is, that, when I gave that wrapper class an empty options array, that class suggested a dynamic control and called the ajax routine. When I gave a filled optionsarray, it suggested a static control and did not call the ajax. So... in 3.5 it worked, because I did the preset of the default using javascript initSelection method. So optionsarray was empty in dynamic select2 controls and filled with data in static controls.
In V4 I set one option even for dynamic select2 controls to have a default; so this wrapper suggested to created a static select2 control in my project...

Arrrghh... sorry sorry sorry. By big fault. Everything is working fine.
Thank you very much for your time and your help, Kevin.

P.S.: Great project!
